px.py

```diff
@@ -715,41 +744,111 @@ def do_CONNECT(self):
         dprint("Entering")
 
         cl = 0
+        cs = 0
         resp, headers, body = self.do_transaction()
         if resp >= 400:
             dprint("Error %d" % resp)
             self.send_error(resp)
         else:
-            dprint("Tunneling through proxy")
-            self.send_response(200, "Connection established")
-            self.send_header("Proxy-Agent", self.version_string())
-            self.end_headers()
-
-            rlist = [self.connection, self.proxy_socket]
-            wlist = []
-            count = 0
-            max_idle = State.config.getint("settings", "idle")
-            while not State.exit:
-                count += 1
-                (ins, _, exs) = select.select(rlist, wlist, rlist, 1)
-                if exs:
-                    break
-                if ins:
-                    for i in ins:
-                        if i is self.proxy_socket:
-                            out = self.connection
-                        else:
-                            out = self.proxy_socket
-
-                        data = i.recv(4096)
-                        if data:
-                            out.send(data)
-                            count = 0
-                            cl += len(data)
-                if count == max_idle:
-                    break
-
-        dprint("Transferred %d bytes" % cl)
+            # Proxy connection may be already closed due to header (Proxy-)Connection: close
+            # received from proxy -> forward this to the client
+            if self.proxy_socket is None:
+                dprint("Proxy connection closed")
+                self.send_response(200, "True")
+                self.send_header("Proxy-Connection", "close")
+                self.end_headers()
```

Comment:
Who is setting `self.proxy_socket = None` when connection is closed?

Reply:
I see you are further down but have separate comment on that.

Reply:
It's at the end of do_socket() if the target proxy sends a "(Proxy-)Connection: close" header (not new code).

Reply:
Shouldn't we instead open another connection with the proxy instead?

Reply:
No, the proxy closes the connection for a reason. Px is only man in the middle and should forward this to the client. The client is responsible for initiating another connect.

Reply:
We are the client as far as the NTLM proxy is concerned. We can send it back to our client but that's another round trip that will come to us anyway.

Regardless, we can optimize this later since it isn't related to this PR.

Reply:
If the proxy closes the connection it's very likely that it can't satisfy the request and every further try will get the same result. I got this on a misconfigured PAC file when px was directed to the wrong proxy which denies a CONNECT for the target URL with a response code 200 and a "Proxy-Connection: close" header. For me forwarding this response to the client seems the only way to handle this. Px can't fix the problem but the client user can (probably).